File: editor/src/de/project/ice/editor/editors/ValueEditor.java
package de.project.ice.editor.editors;


import com.badlogic.gdx.utils.Align;

import java.lang.reflect.Field;

public class ValueEditor<T> extends BaseEditor
{
    private String title;
    protected T value;
    private T oldValue;
    protected Field field;

    protected Object target;

    @Override
    public void act(float delta)
    {
        super.act(delta);

        if ((oldValue != null && !oldValue.equals(value)) || (oldValue == null && value != null))
        {
            oldValue = value;
            setValue(field, target, value, setter);
            fireValueChanged();
        }
        else
        {
            try
            {
                if (!value.equals(getValue(field, target, getter)))
                {
                    value = (T) getValue(field, target, getter);
                    updateValue();
                }
            }
            catch (NullPointerException ignore)
            {
            }
        }
    }

    protected void updateValue()
    {

    }

    @Override
    public ValueEditor<T> bind(Field field, Object target)
    {
        this.target = target;
        this.field = field;
        this.field.setAccessible(true);
        try
        {
            this.value = (T) field.get(target);
        }
        catch (IllegalAccessException e)
        {
            e.printStackTrace();
        }
        this.oldValue = value;
        this.title = field.getName() + ": ";

        defaults().align(Align.left);
        add(title);
        createUi();
        return this;
    }

    protected void createUi()
    {

    }

    protected void fireValueChanged()
    {
        try
        {
            T t = value;
            field.set(target, t);
        }
        catch (IllegalAccessException e)
        {
            e.printStackTrace();
        }
    }
}

File: editor/src/de/project/ice/editor/editors/ValueEditor.kt
package de.project.ice.editor.editors


import java.lang.reflect.Field

open class ValueEditor<T> : BaseEditor() {
    private var title: String? = null
    protected var value: T? = null
    private var oldValue: T? = null
    protected var field: Field? = null
    protected var target: Any? = null
    protected var fieldDescription: String = ""
        private set

    override fun act(delta: Float) {
        super.act(delta)

        if ((oldValue != null && oldValue != value) || (oldValue == null && value != null)) {
            oldValue = value
            BaseEditor.Companion.setValue(field!!, target!!, value!!, setter)
            fireValueChanged()
        } else {
            try {
                if (value != BaseEditor.Companion.getValue(field!!, target!!, getter)) {
                    value = BaseEditor.Companion.getValue(field!!, target!!, getter) as T?
                    updateValue()
                }
            } catch (ignore: NullPointerException) {
            }

        }
    }

    protected open fun updateValue() {

    }

    override fun bind(field: Field, target: Any, description: String?): ValueEditor<T> {
        this.target = target
        this.field = field
        this.fieldDescription = description ?: ""
        field.isAccessible = true
        try {
            this.value = field.get(target) as T
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }

        this.oldValue = value
        this.title = field.name + ": "

        defaults().align(Align.left)
        val titleLabel = VisLabel(title)
        if (description != null) {
            Tooltip.Builder(description).target(titleLabel).build()
        }
        add(titleLabel)
        createUi()
        return this
    }

    protected open fun createUi() {

    }

    protected fun fireValueChanged() {
        try {
            val t = value
            field?.set(target, t)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }

    }
}
